src/v/storage/ntp_config.h

```diff
@@ -15,18 +15,64 @@
 #include "model/metadata.h"
 #include "model/namespace.h"
 #include "ssx/sformat.h"
+#include "storage/file_sanitizer_types.h"
 #include "utils/tristate.h"
 
 #include <seastar/core/sstring.hh>
 
+#include <fmt/format.h>
+
 #include <optional>
+#include <ostream>
 
 namespace storage {
 using with_cache = ss::bool_class<struct log_cache_tag>;
 using topic_recovery_enabled
   = ss::bool_class<struct topic_recovery_enabled_tag>;
 
+/// This represents a set of capabilities that a topic can have.
+class ntp_caps {
+public:
+    explicit ntp_caps(
+      bool tiered_storage, bool iceberg, bool cloud_topics) noexcept
+      : _tiered_storage(tiered_storage)
+      , _iceberg(iceberg)
+      , _cloud_topics(cloud_topics) {}
+
+    /// Tiered storage can be enabled for this NTP
+    bool has_tiered_storage() const noexcept { return _tiered_storage; }
+    /// Iceberg can be enabled for this NTP
+    bool has_iceberg() const noexcept { return _iceberg; }
+    /// Cloud topics can be enabled for this NTP
+    bool has_cloud_topics() const noexcept { return _cloud_topics; }
+
+private:
+    bool _tiered_storage;
+    bool _iceberg;
+    bool _cloud_topics;
+};
+
 class ntp_config {
+    // Make caps based on ntp.
+    // For now we're disallowing Tiered Storage, Iceberg and Cloud Topics
+    // for internal topics and consumer offsets.
+    // The cloud topics are disabled if there is no specific override set
+    // during the topic creation because cloud topic can't be transformed
+    // into non-cloud topic and vice versa.
+    static ntp_caps
+    make_caps(const model::ntp& ntp, bool cloud_topics_override = false) {
+        bool is_internal_topic = ntp.ns == model::kafka_internal_namespace;
+        bool is_consumer_offsets = ntp.ns == model::kafka_namespace
+                                   && ntp.tp.topic
+                                        == model::kafka_consumer_offsets_topic;
+        // For now only take into account the namespace and topic name.
+        return ntp_caps(
+          !is_consumer_offsets && !is_internal_topic,
+          !is_consumer_offsets && !is_internal_topic,
+          // Once set cloud-topics override can not be changed.
+          !is_consumer_offsets && !is_internal_topic && cloud_topics_override);
```

Comment:
nit: this is fine, though I'm not particularly sold on precomputing and encapsulating it in a class, vs a set of utilities functions in a single location that are like `name_supportes_tiered_storage(ntp)` and `name_supports_iceberg(ntp)` or somesuch. I found the separation of the make_caps() impl and the ntp_caps class itself a bit awkward to read

Reply:
I want to be able to print this thing and pass it around.